src/arith/conjunctive_normal_form.cc

```diff
@@ -0,0 +1,415 @@
+    return it->second;
+  }
+
+  Key key{expr_to_key.size()};
+  expr_to_key[expr] = key;
+  key_to_expr[key] = expr;
+  return key;
+}
+
+PrimExpr AndOfOrs::GetExpr(AndOfOrs::Key key) const {
+  auto it = key_to_expr.find(key);
+  ICHECK(it != key_to_expr.end());
+  return it->second;
+}
+
+PrimExpr AndOfOrs::AsPrimExpr() const {
+  PrimExpr expr = Bool(true);
+  for (const auto& chunk : chunks) {
+    PrimExpr chunk_expr = Bool(false);
+    for (Key j : chunk) {
+      chunk_expr = chunk_expr || GetExpr(j);
+    }
+    expr = expr && chunk_expr;
+  }
+  return expr;
+}
+
+void AndOfOrs::TrySimplifyOr(Key* a_ptr, Key* b_ptr, Analyzer* analyzer) {
+  Key& a = *a_ptr;
+  Key& b = *b_ptr;
+  PrimExpr joint = GetExpr(a) || GetExpr(b);
+  PrimExpr simplified = analyzer->Simplify(joint);
+  if (!ExprDeepEqual()(simplified, joint)) {
+    if (auto* simplified_or = simplified.as<OrNode>()) {
+      a = GetKey(simplified_or->a);
+      b = GetKey(simplified_or->b);
+    } else {
+      a = key_false;
+      b = GetKey(simplified);
+    }
+  }
+}
+
+void AndOfOrs::TrySimplifyAnd(Key* a_ptr, Key* b_ptr, Analyzer* analyzer) {
+  Key& a = *a_ptr;
+  Key& b = *b_ptr;
+  PrimExpr joint = GetExpr(a) && GetExpr(b);
+  PrimExpr simplified = analyzer->Simplify(joint);
+  if (!ExprDeepEqual()(simplified, joint)) {
+    if (auto* simplified_and = simplified.as<AndNode>()) {
+      a = GetKey(simplified_and->a);
+      b = GetKey(simplified_and->b);
+    } else {
+      a = key_true;
+      b = GetKey(simplified);
+    }
+  }
+}
+
+void AndOfOrs::Simplify(Analyzer* analyzer) {
+  SimplifyWithinChunks(analyzer);
+  Cleanup();
+  SimplifyAcrossChunks(analyzer);
+  Cleanup();
+}
+
+void AndOfOrs::SimplifyWithinChunks(Analyzer* analyzer) {
+  for (auto& chunk : chunks) {
+    for (size_t expr_i = 0; expr_i < chunk.size(); expr_i++) {
+      for (size_t expr_j = expr_i + 1; expr_j < chunk.size(); expr_j++) {
+        Key& key_i = chunk[expr_i];
+        Key& key_j = chunk[expr_j];
+
+        TrySimplifyOr(&key_i, &key_j, analyzer);
+      }
+    }
+  }
+}
+
+void AndOfOrs::SimplifyAcrossChunks(Analyzer* analyzer) {
+  for (size_t i_and = 0; i_and < chunks.size(); i_and++) {
+    for (size_t j_and = i_and + 1; j_and < chunks.size(); j_and++) {
+      auto& i_chunk = chunks[i_and];
+      auto& j_chunk = chunks[j_and];
+
+      if (i_chunk.size() == 1 && j_chunk.size() == 1) {
+        auto& key_i = i_chunk[0];
+        auto& key_j = j_chunk[0];
+        TrySimplifyAnd(&key_i, &key_j, analyzer);
+        continue;
+      }
+      std::unordered_set<Key> j_set(j_chunk.begin(), j_chunk.end());
+
+      std::optional<size_t> i_distinct_index;
+      for (size_t i = 0; i < i_chunk.size(); i++) {
+        if (!j_set.count(i_chunk[i])) {
+          i_distinct_index = i;
+          break;
+        }
+      }
+
+      if (!i_distinct_index.has_value()) {
+        // I = (i_0 || i_1 || ... || i_N)
+        // J = (i_0 || i_1 || ... || i_N || j_0 || ... || j_N)
```

Comment:
Didn't quite follow this, why is having one distinct index match in j_set sufficient to assume J is a superset?

Reply:
The check is the other way around, that there is no distinct index in i.  That is, `i_distinct_index` is not holding a value, so every element in `i_chunk` is also inside `j_chunk`.